lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/NetworkSpec.hs

```diff
@@ -0,0 +1,384 @@
+-- chain heights.
+mkBlockHeaderHeights
+    :: [BlockHeader]
+    -> [(Hash "BlockHeader", (BlockHeader, Quantity "block" Natural))]
+mkBlockHeaderHeights nodeChain =
+    [ (hash, (hdr, height))
+    | ((hash, hdr), height) <- zip (headerIds nodeChain) [Quantity 1..] ]
+
+{-------------------------------------------------------------------------------
+                              Test chain functions
+-------------------------------------------------------------------------------}
+
+-- | Create a mapping from test chain block ids to block headers.
+headerIds :: [BlockHeader] -> [(Hash "BlockHeader", BlockHeader)]
+headerIds bs = [(prevBlockHash b, a) | (a, b) <- zip bs (tail bs)]
+
+-- | Tip of a test chain is the penultimate block.
+tipId :: [BlockHeader] -> Maybe (Hash "BlockHeader")
+tipId bs = prevBlockHash <$> lastMay bs
+
+-- | A test chain needs at least two headers to have a tip.
+hasTip :: [BlockHeader] -> Bool
+hasTip = (> 1) . length
+
+-- | Length of a test chain, not including the block header after the tip.
+chainLength :: [BlockHeader] -> Int
+chainLength bs = max 0 (length bs - 1)
+
+chainTip :: [BlockHeader] -> Maybe BlockHeader
+chainTip = initMay >=> lastMay
+
+-- | Slot index of the tip of a chain.
+chainEnd :: [BlockHeader] -> SlotId
+chainEnd = maybe (SlotId 0 0) slotId . chainTip
+
+-- | Limit the sequence to a certain size by removing items from the beginning.
+limitChain :: Quantity "block" Natural -> [BlockHeader] -> [BlockHeader]
+limitChain (Quantity k) bs = drop (max 0 (length bs - fromIntegral k - 1)) bs
+
+showChain :: [BlockHeader] -> String
+showChain [] = ""
+showChain bs = mconcat
+    [ showSlot (slotId (head bs)), " | "
+    , unwords (map showHeaderHash (tail bs)) ]
+  where
+    showHeaderHash (BlockHeader _ (Hash h)) = B8.unpack h
+
+showSlot :: SlotId -> String
+showSlot (SlotId ep sl) = show ep ++ "." ++ show sl
+
+{-------------------------------------------------------------------------------
+                     Test chain pure model of intersection
+-------------------------------------------------------------------------------}
+
+-- | Merge node chain with local unstable blocks.
+modelIntersection
+    :: Quantity "block" Natural
+    -- ^ Maximum number of unstable blocks.
+    -> [BlockHeader]
+    -- ^ Local test chain.
+    -> [BlockHeader]
+    -- ^ Remote test chain.
+    -> [BlockHeader]
+    -- ^ New local test chain.
+modelIntersection k localChain nodeChain =
+    -- We have at most k block headers ...
+    limitChain k $
+    -- ... added to the end of current unstable block headers ...
+    maybe id dropToSlot (min p (p >>= fixup))
+    -- ... fetched from the node.
+    nodeChain
+  where
+    p = intersectionPoint localChain nodeChain
+    -- The real code under test will fetch the full k if the local chain is
+    -- ahead of the node chain. It's not incorrect behaviour, and the situation
+    -- should never happen anyway. But we need to add this to the model so that
```

Comment:
There is actually one case where I'd see this happening: when reconnecting a wallet after a long-enough interruption with the network. Especially if we don't persist the unstable blocks to disk. We'll originally start with an empty list of unstable blocks and, upon reconnecting, we'll have to populate it all. Now, if we get disconnected from the node for let's say 2161 blocks, then our whole local chain is likely outdated regardless of whether there has been roll backs or not and we'd have to fetch `k` blocks again. 

Do you agree?